### PR TITLE
raftstore: add entry cache

### PR DESCRIPTION
**Pull request by BusyJay**
Add cache for fetching log entries, this can reduce the time visiting rocksdb.

### CODE REVIEW

**Comment by siddontang** [comment]
I think it is better to abstract an EentryCache separately which you can test it directly and make the PeetStorage more simple.

**Comment by BusyJay** [comment]
Event I abstract an `EntryCache`, I still need to test if some temporary variable is computed correctly and ensure cache's content is the same as rocksdb. Hence I still need to test with `PeerStorage`.

**Comment by siddontang** [comment]
PTAL @javaforfun @hhkbp2

**Comment by javaforfun** [comment]
seems only limit cache number?   for one region, max memory usage will be MAX_CACHE_CAPACITY * max_entry_size = 3K * 16K = 48M.
so for 1K region,  max cache memory could be 48G?

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -316,10 +323,61 @@ impl PeerStorage {
         if low == high {
             return Ok(ents);
         }
+        let cache_low = self.cache.front().map_or(0, |e| e.get_index());
+        if self.cache.is_empty() || high <= cache_low {
+            // not overlap
```
Comment (hhkbp2) [inline src/raftstore/store/peer_storage.rs]: It's better to wrap all the cache into a standalone struct or file.

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -316,10 +323,61 @@ impl PeerStorage {
...
+            if fetched_size > max_size {
+                // max_size exceed.
+                return Ok(ents);
+            }
+            0
+        } else {
+            (low - cache_low) as usize
+        };
+        let mut end_idx = begin_idx;
+        self.cache
```
Comment (hhkbp2) [inline src/raftstore/store/peer_storage.rs]: add metrics or log for the cache hit rate.

File: src/raftstore/store/peer_storage.rs [hunk 3]
```diff
@@ -637,6 +747,7 @@ impl PeerStorage {
                                      raft_count += 1;
                                      Ok(true)
                                  }));
+        self.cache = VecDeque::new();
```
Comment (hhkbp2) [inline src/raftstore/store/peer_storage.rs]: Preallocation maybe?

**Comment by BusyJay** [comment]
Yes, if they are all active. Once some region is not active anymore, it's cache [will be free](https://github.com/pingcap/tikv/pull/1984/files#diff-ac5fa05712c478a85585b4852858d036R672).

File: src/raftstore/store/peer_storage.rs (hunk 3 above)
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Not necessary. If the region is inactive, preallocation will be a waste. We usually have many regions which is a big cost.

**Comment by siddontang** [comment]
PTAL @hhkbp2 @javaforfun

**Comment by javaforfun** [comment]
I think it's necessary to limit max cache memory,  rest LGTM.

**Comment by BusyJay** [comment]
Maybe next pr.

File: src/raftstore/store/peer_storage.rs [hunk 4]
```diff
@@ -44,6 +46,10 @@ pub const RAFT_INIT_LOG_INDEX: u64 = 5;
 const MAX_SNAP_TRY_CNT: usize = 5;
 const RAFT_LOG_MULTI_GET_CNT: u64 = 8;
 
+// One extra slot for VecDeque internal usage.
+const MAX_CACHE_CAPACITY: usize = 1024 * 3 - 1;
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: the max value is too large here?

Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: What's your favorite number?

Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: The default max inflight msgs is 256, I guess maybe 1024 is enough.

File: src/raftstore/store/peer_storage.rs [hunk 5]
```diff
@@ -104,6 +110,131 @@ pub fn first_index(state: &RaftApplyState) -> u64 {
     state.get_truncated_state().get_index() + 1
 }
 
+#[derive(Default)]
+struct EntryCache {
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: do we have any test to test EntryCache directly now?

Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: No, all tests are tested via `PeerStorage`, but all `EntryCache`'s methods are covered.

**Comment by hhkbp2** [comment]
LGTM

**Comment by BusyJay** [comment]
According to bench, CPU drop 4% ~ 2%, not as large as we thought.

**Comment by siddontang** [comment]
It is still worth to support it now. Without cache, we many use the iterator to retrieve raft logs, which reduces the performance.

**Comment by hhkbp2** [comment]
> According to bench, CPU drop 4% ~ 2%

Similar to some test previously run.

**Pull request merged**